### DISCUSSION TITLE
Babel transformer to automatically insert makeObservable into class constructor when mobx decorators are used

### DISCUSSION THREAD

**Discussion by olee** [mobxjs/mobx#3314] [Ideas]
As the title says, I think it would be a nice feature to write a babel decorator to automate adding `makeObservable` to constructurs.
It could work like this:
- Check, if a class uses mobx decorators, otherwise bail
- Check, if the constructor already invokes `makeObservable` (or variants), otherwise bail
- Automatically inject an invocation of `makeObservable(this)` into the constructor (and perhaps create a default constructor implementation if the class does not yet have one)

I think this should be fairly simple to implement as a transformer and could make writing mobx code when using decorators as easy and straight-forward like when using mobx 4/5.

Another alternative I was thinking of would be writing a decorator which wraps the constructor call to automatically invoke `makeObservable`, but I think that might be error prone, because invoking `makeObservable` before the default ctor would probably not work and afterwards it might cause issues if observable fields are used in the constructor code, so I scrapped that idea.

**Comment by urugator** [comment]
Have you checked https://github.com/mobxjs/mobx/tree/main/packages/eslint-plugin-mobx#mobxmissing-make-observable ?

**Reply by olee** [reply]
That looks awesome already.
What I suggest would basically be a babel plugin which does this automatically when building 👍

**Reply by urugator** [reply]
I've been thinking about using transpiler, but mobx is designed to do everything at runtime (which also implies some usage constraints). Once you introduce a compilation step a lot of interesting possibilities opens up and you have to ask yourself whether it still does make sense to use mobx. It's not wrong, but to me it feels like a bandate that introduces another layer of complexity and does quite little ...
Also I think it's something that can be implemented outside the mobx package and can be given a mention in documentation if appropriate.

**Reply by olee** [reply]
Sure, I never thought of this as something which has to be directly part of the mobx package, though if it is some really small transformer, it maybe could be added to the package as well as something like `mobx/makeObservable-transformer`.
But primarily I'd see this as a separate package which could be used alongside mobx only if one wants to.
